beacon-chain/forkchoice/doubly-linked-tree/unrealized_justification.go

```diff
@@ -0,0 +1,51 @@
+package doublylinkedtree
+
+import types "github.com/prysmaticlabs/prysm/consensus-types/primitives"
+
+func (s *Store) setUnrealizedJustifiedEpoch(root [32]byte, epoch types.Epoch) error {
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+
+	node, ok := s.nodeByRoot[root]
+	if !ok || node == nil {
+		return ErrNilNode
+	}
+	if epoch < node.unrealizedJustifiedEpoch {
+		return errInvalidUnrealizedJustifiedEpoch
+	}
+	node.unrealizedJustifiedEpoch = epoch
+	return nil
+}
+
+func (s *Store) setUnrealizedFinalizedEpoch(root [32]byte, epoch types.Epoch) error {
+	s.nodesLock.Lock()
+	defer s.nodesLock.Unlock()
+
+	node, ok := s.nodeByRoot[root]
+	if !ok || node == nil {
+		return ErrNilNode
+	}
+	if epoch < node.unrealizedFinalizedEpoch {
+		return errInvalidUnrealizedFinalizedEpoch
+	}
+	node.unrealizedFinalizedEpoch = epoch
+	return nil
+}
+
+// UpdateUnrealizedCheckpoints "realizes" the unrealized justified and finalized
+// epochs stored within nodes. It should be called at the beginning of each
+// epoch
+func (f *ForkChoice) UpdateUnrealizedCheckpoints() {
+	f.store.nodesLock.Lock()
+	defer f.store.nodesLock.Unlock()
+	for _, node := range f.store.nodeByRoot {
```

Comment:
Maybe a silly question, do we need to do this for all the nodes? or only the tips?

Reply:
We need only the tips but the cost of finding them involves a full loop anyway so it's better to just update then all.